internals/cli/app.go

```diff
@@ -0,0 +1,371 @@
+				err := flag.flag.Value.Set(os.Getenv(flag.envVar))
+				if err != nil {
+					return err
+				}
+			}
+		}
+		for _, flag := range c.flags {
+			if !flag.flag.Changed && flag.HasEnvarValue() {
+				err := flag.flag.Value.Set(os.Getenv(flag.envVar))
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+}
+
+// Flag defines a new flag with the given long name and help text,
+// adding an environment variable default configurable by APP_COMMAND_FLAG_NAME.
+// The help text is suffixed with the default value of the flag.
+func (c *CommandClause) Flag(name string) *Flag {
+	if _, ok := c.flags[name]; ok {
+		return c.flags[name]
+	}
+	fullCmd := strings.Replace(c.fullCommand(), " ", c.App.separator, -1)
+	if cmd, ok := c.isPersistentFlag(name); ok {
+		fullCmd = strings.Replace(cmd.fullCommand(), " ", c.App.separator, -1)
+	}
+	prefix := formatName(fullCmd, "", c.App.separator, c.App.delimiters...)
+	envVar := formatName(name, prefix, c.App.separator, c.App.delimiters...)
+
+	c.App.registerEnvVar(envVar)
+	baseFlag := c.Cmd.Flag(name)
+	flag := (&Flag{
+		flag:   baseFlag,
+		app:    c.App,
+		envVar: envVar,
+	}).Envar(envVar)
+	if c.flags == nil {
+		c.registerEnvVarParsing()
+		c.flags = make(map[string]*Flag)
+	}
+	c.flags[name] = flag
+	return flag
+}
+
+func (c *CommandClause) Flags() *FlagSet {
+	return &FlagSet{FlagSet: c.Cmd.Flags(), cmd: c}
+}
+
+func (c *CommandClause) isPersistentFlag(name string) (*CommandClause, bool) {
+	if c.Cmd == c.Cmd.Root() {
+		return nil, false
+	}
+	var parent *CommandClause
+	persistent := false
+	for p := c.Cmd.Parent(); p != nil; p = p.Parent() {
+		f := p.Flags()
+		f.VisitAll(func(flag *pflag.Flag) {
+			if flag.Name == name {
+				persistent = true
+				parent = &CommandClause{Cmd: p}
+			}
+		})
+	}
+	return parent, persistent
+}
+
+// BindArguments binds a function to a command clause, so that
+// it is executed when the command is parsed.
+func (c *CommandClause) BindArguments(params []Argument) {
+	c.Args = params
+	if params != nil {
+		c.Cmd.PreRunE = MergeFuncs(c.Cmd.PreRunE, func(cmd *cobra.Command, args []string) error {
+			if err := c.argumentError(args); err != nil {
+				return err
+			}
+			return ArgumentRegister(params, args)
+		})
+	}
+}
+
+// BindArgumentsArr binds a function to a command clause, so that
+// it is executed when the command is parsed.
+func (c *CommandClause) BindArgumentsArr(params []Argument) {
+	c.Args = params
+	if params != nil {
+		c.Cmd.PreRunE = MergeFuncs(c.Cmd.PreRunE, func(cmd *cobra.Command, args []string) error {
+			if len(args) <= 0 {
+				return c.argumentError(args)
+			}
+			return ArgumentArrRegister(params, args)
+		})
+	}
+}
+
+func (c *CommandClause) BindAction(fn func() error) {
+	if fn != nil {
+		c.Cmd.RunE = func(cmd *cobra.Command, args []string) error {
+			return fn()
+		}
+	}
+}
+
+type PreRunAction func(*cobra.Command, []string) error
+
+func MergeFuncs(f1 PreRunAction, f2 PreRunAction) PreRunAction {
```

Comment:
I know this is code written by me, but I think it would be a bit more clear if instead of always writing:
```go
c.Cmd.PreRunE = MergeFuncs(c.Cmd.PreRunE, ...)
```

We could just write:
```go
c.AddPreRunE(...)
```
to add a pre run function to a command clause.

(Based on a suggestion by Joris)

Reply:
Solved.